api/src/permissions/modules/validate-access/lib/validate-item-access.ts

```diff
@@ -43,22 +55,61 @@ export async function validateItemAccess(options: ValidateItemAccessOptions, con
 		},
 	};
 
+	// Inject the root fields after the permissions have been processed, as to not require access to all collection fields
+	if (options.returnAllowedRootFields) {
+		ast.children = Object.keys(context.schema.collections[options.collection]!.fields).map((field) => ({
+			type: 'field',
+			name: field,
+			fieldKey: field,
+			whenCase: [],
+			alias: false,
+		}));
+
+		const policies = await fetchPolicies(options.accountability, context);
+
+		const permissions = await fetchPermissions(
+			{ action: options.action, policies, collections: [options.collection], accountability: options.accountability },
+			context,
+		);
+
+		injectCases(ast, permissions);
+	}
+
 	const items = await fetchPermittedAstRootFields(ast, {
 		schema: context.schema,
 		accountability: options.accountability,
 		knex: context.knex,
 		action: options.action,
 	});
 
-	if (items && items.length === options.primaryKeys.length) {
-		const { fields } = options;
+	const hasAccess = items && items.length === options.primaryKeys.length;
+
+	if (!hasAccess) {
+		return { accessAllowed: false };
+	}
+
+	let accessAllowed = true;
+
+	// If specific fields were requested, verify they are all accessible
+	if (options.fields) {
+		accessAllowed = items.every((item: any) => options.fields!.every((field) => toBoolean(item[field])));
+	}
+
+	// If returnAllowedRootFields, build a map of allowed fields per item
+	if (options.returnAllowedRootFields && items.length > 0) {
+		const allowedRootFieldsMap: Record<string, string[]> = {};
+
+		for (const item of items) {
+			const id = String(item[primaryKeyField]);
```

Comment:
Wouldn’t this end up being a 0 or 1 rather than the actual value? I don’t think we can reliably determine which record corresponds to which item/field permissions. We may need to remove the map and handle it the same way we process fields: if a field has a value of 1 for all items, then it can be considered allowed for all items and should be marked as such.